### ISSUE TITLE
[MPS] Consecutive calls to F.linear produce different results for >2D inputs

### ISSUE THREAD

**Issue by tobiasofsn** [pytorch/pytorch#180776]
### 🐛 Describe the bug

Calling `F.linear(x, w, b)` multiple times with the same input produces different results.

I have only been able to reproduce it when the following conditions were met:
- Backend is MPS
- Data type is `BFloat16` or `Float16`
- Bias `b` is `None`
- Input `x` has more than 2 dimensions
- Input `x` has one of the affected shapes

### Context

I originally encountered this issue in an LLM fine-tuning workflow using `transformers` and `trl`, where training with `DPOTrainer` would result in corrupted training losses, and `GRPOTrainer` would generate nonsensical text during training. I traced the root cause to `nn.Linear`(`F.linear`). Without the workaround below, I have to resort to `Float32`.

### MRE

```python
import torch
import torch.nn.functional as F

x = torch.randn(2, 13, 1024, device="mps", dtype=torch.bfloat16)
w = torch.randn(1024, 1024, device="mps", dtype=torch.bfloat16)

first = F.linear(x, w).clone()
second = F.linear(x, w).clone()

print("Diff between runs:", (second - first).abs().max().item())
```
Output:
`Diff between runs: 123.5`

Here are a few examples of other shapes that are affected:

- x: `(6, 6, 634)`, w: `(634, 634)`
- x: `(1, 3, 28, 315)`, w: `(315, 315)`
- x: `(1, 12, 4, 512)`, w: `(512, 512)`
- x: `(1, 1, 5, 6, 1024)`, w: `(1024, 1024)`

### Workaround

I cannot reproduce the issue after applying this patch, which reshapes the input to 2D, calls the original F.linear, and reshapes it back afterwards.

```python
original_linear = F.linear
def patched_linear(input, weight, bias=None):
    if input.device.type == 'mps' and (input.dtype == torch.bfloat16 or input.dtype == torch.float16) and bias is None and input.dim() > 2:
        *leading, h = input.shape
        return original_linear(input.reshape(-1, h), weight, bias).reshape((*leading, weight.shape[0]))
    return original_linear(input, weight, bias)
F.linear = patched_linear
```

### Possibly related

#178203 - Detect mm padding overflow and incorrect alignment conditions and dispatch to metal_mm

@malfet @kulinseth @albanD @razarmehr @jhavukainen 

### Versions

PyTorch version: 2.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 26.4.1 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.0.123.102)
CMake version: Could not collect
Libc version: N/A

Python version: 3.14.3 (v3.14.3:323c59a5e34, Feb  3 2026, 11:41:37) [Clang 16.0.0 (clang-1600.0.26.6)] (64-bit runtime)
Python platform: macOS-26.4.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M5

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] torch==2.11.0
[pip3] torchvision==0.26.0
[conda] Could not collect

cc @ezyang @gchanan @kadeng @msaroufim @mruberry @kurtamohler @jianyuh @nikitaved @walterddr @xwang233 @Lezcano @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk

**Comment by malfet** [comment]
Either myself or @jhavukainen did fix very similar bug for bmm: MPS does not understand 0-stride tensors

**Comment by malfet** [comment]
Interestingly enough, it seems HW specific bug - I could not repro it on M4, but it only fails on M5
